### PR TITLE
[Enhancement]adaptive coalesce active column and lazy column

### PR DESCRIPTION
**Pull request by zombee0**
In some cases, active and lazy column will be coalesced together, When requeting io, their io will be requested together, in case of low-pass filter, lazy column is useless, we adaptivly coalesce them.
On each scan node, there is a counter, when lazy_columns are needed counter++ else counter--, of course it's atomic.
when counter>0, so we know lazy_columns are needed has a greater possibility, so we coalesce them with active_columns.

Fixes #issue
a case: 100columns, 1.3M/column/rowgroup, 1/8 row_group need lazy column.
截屏2023-06-26 10 27 38
to 
截屏2023-06-26 10 29 10

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

File: be/src/connector/hive_connector.cpp [hunk 1]
```diff
@@ -482,6 +482,7 @@ Status HiveDataSource::_init_scanner(RuntimeState* state) {
     scanner_params.case_sensitive = _case_sensitive;
     scanner_params.profile = &_profile;
     scanner_params.open_limit = nullptr;
+    scanner_params.lazy_column_colease_counter = get_lazy_column_colease_counter();
```
Comment (dirtysalt) [inline be/src/connector/hive_connector.cpp:485]: coalesce

**Comment by dirtysalt** [comment]
Could you describe a little about , when coalesce lazy columns and when not?

File: be/src/exec/hdfs_scanner.cpp [hunk 2]
```diff
@@ -131,6 +131,7 @@ Status HdfsScanner::_build_scanner_context() {
     ctx.timezone = _runtime_state->timezone();
     ctx.iceberg_schema = _scanner_params.iceberg_schema;
     ctx.stats = &_stats;
+    ctx.lazy_column_coalesce_counter = _scanner_params.lazy_column_colease_counter;
```
Comment (DorianZheng) [inline be/src/exec/hdfs_scanner.cpp:134]: :typo
`_scanner_params.lazy_column_coalesce_counter`

**Review by dirtysalt — APPROVED** [review]

File: be/src/io/shared_buffered_input_stream.cpp [hunk 3]
```diff
@@ -94,10 +83,105 @@ Status SharedBufferedInputStream::set_io_ranges(const std::vector<IORange>& rang
...
+            }
+        }
+    }
+
     _update_estimated_mem_usage();
     return Status::OK();
 }
 
+Status SharedBufferedInputStream::set_io_ranges(const std::vector<IORange>& ranges, bool coalesce_together) {
+    if (coalesce_together || !config::io_coalesce_adaptive_lazy_active) {
```
Comment (Youngwb) [inline be/src/io/shared_buffered_input_stream.cpp:178]: if io_coalesce_adaptive_lazy_active set to false, it will coalesce together?

**Review by Youngwb — APPROVED** [review]

**Pull request merged**